[Merged by Bors] - feat(data/mv_polynomial/derivation): derivations of `mv_polynomial`s

urkud opened the pull request:
---

- [x] depends on: #9132
- [x] depends on: #9137 
- [x] depends on: #10945 

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

urkud committed:
Snapshot

urkud committed:
Apply suggestions from code review

urkud committed:
Merge branch 'master' into mv-poly-golf

urkud committed:
feat(data/finsupp): more lemmas about `α →₀ ℕ`

urkud committed:
Merge branch 'finsupp-nat' into mv-poly-deriv

urkud committed:
Snapshot

eric-wieser commented on src/data/mv_polynomial/derivation.lean:
```diff
@@ -0,0 +1,122 @@
+  f = g :=
+begin
+  apply derivation.to_linear_map_injective, ext s, dsimp,
+  induction s using finsupp.induction with i n s hi hn ihs,
```
It worries me that you need this, shouldn't ext be handling this induction for you?

urkud replied:
No, `f ∘ (λ s, monomial s 1)` is not an object known to `ext`.

eric-wieser replied:
Ah, I understand now.

Can you prove this about `add_monoid_algebra` first?
```lean
@[ext] lemma add_monoid_algebra.derivation_ext {G} [add_comm_monoid G] [module (add_monoid_algebra R G) A]
  [is_scalar_tower R (add_monoid_algebra R G) A]
  {f g : derivation R (add_monoid_algebra R G) A}
  (h : ∀ i, f (add_monoid_algebra.of' _ _ i) = g (add_monoid_algebra.of' _ _ i)) :
  f = g :=
begin
  apply derivation.to_linear_map_injective,
  ext s,
  exact h s,
end
```
then this becomes
```lean
begin
  ext s, show f (monomial s 1) = g (monomial s 1),
  induction s using finsupp.induction with i n s hi hn ihs,
  { exact f.map_one_eq_zero.trans g.map_one_eq_zero.symm },
  { simp only [monomial_single_add, ihs, h, derivation.leibniz, derivation.leibniz_pow] }
end
```

urkud committed:
Merge branch 'master' into mv-poly-deriv

eric-wieser commented on src/data/mv_polynomial/derivation.lean:
```diff
@@ -0,0 +1,122 @@
+variable (R)
+
+def mk_derivationₗ (f : σ → A) :
+  mv_polynomial σ R →ₗ[R] A :=
```
```suggestion
def mk_derivationₗ (f : σ → A) : mv_polynomial σ R →ₗ[R] A :=
```
suggesting +committing this only because the cache for this PR seems corrupt, and I need to change something to force it to rebuild

eric-wieser committed:
Update src/data/mv_polynomial/derivation.lean

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
chore(ring_theory/derivation): add `leibniz_pow`

urkud committed:
Squeeze simp, add map_inv_of

urkud committed:
Merge branch 'YK-leibniz-pow' into mv-poly-deriv

eric-wieser committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
refactor(data/{mv_,}polynomial): lemmas about `adjoin`

Prove `adjoin {X} = ⊤` and `adjoin (range X) = ⊤` for `polynomial`s
and `mv_polynomial`s much earlier and use these equalities to golf
some proofs.

Also drop some `comm_` in typeclass assumptions.

urkud committed:
Fix

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
Merge branch 'YK-adjoin-X' into mv-poly-deriv

urkud committed:
Merge branch 'master' into YK-adjoin-X

urkud committed:
Update

urkud committed:
Fix

urkud committed:
Merge branch 'master' into YK-adjoin-X

urkud committed:
Fix

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
Merge branch 'YK-adjoin-X' into mv-poly-deriv

urkud committed:
Fix

urkud committed:
Merge branch 'mv-poly-deriv' of git://github.com/leanprover-community/mathlib into mv-poly-deriv

urkud committed:
Fix

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
Fix lint

urkud committed:
newlines

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
Update

urkud committed:
Snapshot

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
Merge branch 'master' into mv-poly-deriv

urkud committed:
Revert

urkud committed:
Fix

urkud committed:
Typo

jcommelin commented on src/data/mv_polynomial/basic.lean:
```diff
@@ -125,6 +125,10 @@ instance [monoid R] [comm_semiring S₁] [distrib_mul_action R S₁] [distrib_mu
 add_monoid_algebra.is_central_scalar
 instance [comm_semiring R] [comm_semiring S₁] [algebra R S₁] : algebra R (mv_polynomial σ S₁) :=
 add_monoid_algebra.algebra
+-- Register with high priority to avoid timeout in `data.mv_polynomial.pderiv`
+instance is_scalar_tower' [comm_semiring R] [comm_semiring S₁] [algebra R S₁] :
+  is_scalar_tower R (mv_polynomial σ S₁) (mv_polynomial σ S₁) :=
+is_scalar_tower.right
```
Why do we need this instance? Do we understand the bug?

urkud replied:
As far as I understand, Lean tries to use https://leanprover-community.github.io/mathlib_docs/data/mv_polynomial/basic.html#mv_polynomial.is_scalar_tower and fails.

jcommelin commented on src/data/mv_polynomial/derivation.lean:
```diff
@@ -0,0 +1,141 @@
+/-!
+# Derivations of multivariate polynomials
+
+In this file we prove that a derivation of `mv_polynomial σ R` is defined by its values on all
```
```suggestion
In this file we prove that a derivation of `mv_polynomial σ R` is determined by its values on all
```

urkud committed:
Update src/data/mv_polynomial/derivation.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

jcommelin reviewed (approved):
Thanks :tada:

bors merge
